### ISSUE TITLE
Mystery Race Condition

### ISSUE THREAD

**Issue by TheTechromancer** [blacklanternsecurity/bbot#646]
```
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542 scan._finished_init: True
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542 manager.active: True
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542     manager.running: True
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542         manager._task_counter.value: 2
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542         manager._task_counter.tasks:
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542             - emit_event(DNS_NAME("www.billing.evilcorp.com", module=wayback, tags={'distance-1', 'subdomain'})) running for 19 hours, 21 minutes, 54 seconds:
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542             - emit_event(DNS_NAME("billing.evilcorp.com", module=wayback, tags={'distance-1', 'subdomain'})) running for 19 hours, 21 minutes, 38 seconds:
2023-08-01 13:01:16,722 [DEBUG] bbot.scanner scanner.py:542         manager.incoming_event_queue.qsize: 0

...

2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542         dnscommonsrv.finished: False
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542             running: True
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542             tasks:
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542                 - handle_event(DNS_NAME("evilcorp.com", module=TARGET, tags={'aaaa-error', 'target', 'txt-error', 'distance-0', 'ns-error', 'domain', 'cname-error', 'unresolved', 'mx-error', 'in-scope', 'soa-error', 'srv-error', 'a-error'})) running for 19 hours, 21 minutes, 15 seconds:
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542                 - handle_event(DNS_NAME("evilcorp.com", module=TARGET, tags={'aaaa-error', 'target', 'txt-error', 'distance-0', 'ns-error', 'domain', 'cname-error', 'unresolved', 'mx-error', 'in-scope', 'soa-error', 'srv-error', 'a-error'})) running for 19 hours, 20 minutes, 3 seconds:
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542             num_incoming_events: 0
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542             outgoing_event_queue.qsize: 0
```

**Comment by TheTechromancer** [comment]
What we know:

- two `dnscommonsrv` tasks appeared to stall:

```
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542                 - handle_event(DNS_NAME("evilcorp.com", module=TARGET, tags={'aaaa-error', 'target', 'txt-error', 'distance-0', 'ns-error', 'domain', 'cname-error', 'unresolved', 'mx-error', 'in-scope', 'soa-error', 'srv-error', 'a-error'})) running for 19 hours, 21 minutes, 15 seconds:
2023-08-01 13:01:16,723 [DEBUG] bbot.scanner scanner.py:542                 - handle_event(DNS_NAME("evilcorp.com", module=TARGET, tags={'aaaa-error', 'target', 'txt-error', 'distance-0', 'ns-error', 'domain', 'cname-error', 'unresolved', 'mx-error', 'in-scope', 'soa-error', 'srv-error', 'a-error'})) running for 19 hours, 20 minutes, 3 seconds:
```

- all five `dnscommonsrv` workers were still running when the scan was cancelled:
```
$ cat -n schizophrenic_samwise | grep -iE dnscommon.*worker stopped'
6674556	2023-08-01 13:52:22,052 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker stopped
6674608	2023-08-01 13:52:22,055 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker stopped
6674660	2023-08-01 13:52:22,058 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker stopped
6674662	2023-08-01 13:52:22,059 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker stopped
6674714	2023-08-01 13:52:22,062 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker stopped
```
- but only three of them had CancelledError tracebacks:
```
$ cat -n schizophrenic_samwise | grep -iE 'dnscommon.*worker cancel' 
6674557	2023-08-01 13:52:22,053 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker cancelled
6674609	2023-08-01 13:52:22,055 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker cancelled
6674663	2023-08-01 13:52:22,059 [TRACE] bbot.modules.dnscommonsrv logger.py:90 Worker cancelled
```
- of these three, all were waiting on an event from the queue:
```
Traceback (most recent call last):
  File "/root/bbot/bbot/modules/base.py", line 335, in _worker
    event = await self.incoming_event_queue.get()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/queues.py", line 158, in get
    await getter
asyncio.exceptions.CancelledError
```

I'm inclined to think the task counter is not bugged and that it's reporting the remaining two tasks correctly. That would mean that something else is causing them to stall. I suspect `as_completed()`.

Today I performed heavy fuzzing on `resolve_batch()`, `acatch`, `TaskCounter`, and `RateLimiter`, and `as_completed()`, and all of them seem to be performing as expected even under extremely high load. I even tried fuzzing dnspython with fully random data, and it was rock solid, no errors.

Since it's still unclear what or where the bug is, these are the steps taken:
- add log statements at start and end of every task
- add log statements at start and finish of every `handle_event()`
- add blanket error logging to `dns.resolve()` and `dns.resolve_raw()`
- add locks to `TaskCounter` (shouldn't be necessary but it doesn't hurt)
- disable get_size() in case it is causing unwanted bugs
- included asyncio in BBOT log handler

**Comment by TheTechromancer** [comment]
This is an upstream bug in dnspython. It is reproducible with the following code:

```python
import asyncio
import dns.asyncresolver
import dns.exception

def make_resolver():
    resolver = dns.asyncresolver.Resolver()
    resolver.lifetime = .01
    resolver.timeout = .01
    resolver.nameservers = ["127.0.0.1"]
    return resolver

async def resolve(thing):
    print(f"resolving {thing}")
    try:
        return await make_resolver().resolve(thing, rdtype="A")
    except dns.exception.DNSException:
        pass
    finally:
        print(f"finished resolving {thing}")

async def main():
    batch_size = 100
    while 1:
        names = [f"asdf{i}.com" for i in range(batch_size)]
        tasks = [asyncio.create_task(resolve(n)) for n in names]
        await asyncio.gather(*tasks)

asyncio.run(main())
```

This is being tracked in an upstream issue: https://github.com/rthalley/dnspython/issues/976

**Comment by TheTechromancer** [comment]
Fixed in aec7c4b7aa77ed4ab425789a829a7af7afbabbfe, PR: https://github.com/blacklanternsecurity/bbot/pull/637. We are using `asyncio.wait_for()` to forcefully abort DNS queries if they exceed their timeouts.

**Comment by TheTechromancer** [comment]
From the creator of dnspython:

> After more fiddling I reproduced it on linux. I had to increase the load a bit. The really simple way to reproduce it is
> 
> ```
> #!/usr/bin/env python3
> 
> import asyncio
> 
> import dns.message
> import dns.asyncquery
> 
> async def qasync():
>     q = dns.message.make_query("www.dnspython.org", "A")
>     print(await dns.asyncquery.udp(q, "10.0.0.1", timeout=0.0))
> 
> asyncio.run(qasync())
> ```
> 
> where "10.0.0.1" is some address that will never reply.
> 
> The problem was that in a few places we said "if timeout:" when we really wanted to say "if timeout is not None:" and thus in the case of a zero timeout we fell into the "wait forever" code. If a packet was also lost, then this would hang the async task forever.
> 
> Fixing the bad "if timeout" in the dnspython asyncio backend fixed the problem. I also fixed a few other instances of this mistake I found doing this after a code audit.
> 
> The dnspython main line and also the 2.4 branch have been fixed. I will probably do a 2.4.2 release in a week or two as we've now got two release-worthy bug fixes pending. In the meantime, using a dnspython installed from the source should help if you're in a rush, or you can just apply the change in the commit to your installed dnspython.
